File: solutions/1457/main.py
from typing import * 

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def count_ones(n): 
    res = 0 
    while n: 
        n &= (n-1)
        res += 1 
    return res 

class Solution:
    def dfs(self, node, path):
        # check if reach leaf  
        if node.left is None and node.right is None: 
            if count_ones(path) <= 1: 
                self.res += 1 
            return 
        # 1 <= Node.val <= 9 表示可以用状态压缩
        # 注意：只有在递归的时候才改变path的值
        if node.left is not None: 
            # record left node val 
            self.dfs(node.left, path ^ (1 << node.left.val))
        if node.right is not None: 
            # record right node val 
            self.dfs(node.right, path ^ (1 << node.right.val))
        
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        self.res = 0 
        if root is None: 
            return self.res 
        self.dfs(root, 1 << root.val) 
        return self.res 
        

File: solutions/1457/main.rs
use std::rc::Rc; 
use std::cell::RefCell; 

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}

// use std::rc::Rc;
// use std::cell::RefCell;

// fn main() {
//     let data = Rc::new(RefCell::new(42));

//     // Cloning an Rc increments the reference count
//     let reference1 = Rc::clone(&data);
//     let reference2 = Rc::clone(&data);

//     // Borrowing mutably through RefCell
//     let mut mutable_reference = data.borrow_mut();
//     *mutable_reference += 1;

//     // Since RefCell allows interior mutability, even though data is Rc,
//     // the value can be mutated through the mutable reference
// }

impl Solution {
    pub fn pseudo_palindromic_paths(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(node: Option<&Rc<RefCell<TreeNode>>>, mut mask: i32) -> i32 {
            let check_mask = |mask| return if (mask & (mask - 1)) == 0 { 1 } else { 0 }; 
            if let Some(node) = node {
                let node = node.borrow(); 
                mask ^= 1 << node.val; 
                if node.left.is_none() && node.right.is_none() {
                    return check_mask(mask);
                }
                return dfs(node.left.as_ref(), mask) + dfs(node.right.as_ref(), mask);
            }
            0
        }
        dfs(root.as_ref(), 0)
    }
}